README.md
 
 ## Usage 
You’ll need a server to be able to fetch with import, as it doesn’t work on the file:// protocol.  
[ [https://www.sitepoint.com/using-es-modules/]() ]

 ## Test
 `npm test`

CHANGELOG.md
## 0.3 (2021-05-31)
- Add a way to import logs

## 0.2.1 (2021-05-31)
- Fix export string not updating on rendering

## 0.2 (2021-05-31)
- Add tests
- Add a way to export logs
- Refactor the codebase

## 0.1 (2018-04-30)
- First version

test/test.js
import {strict as assert} from 'assert';
import { JSONDate, Counter, DateCounter, Logger } from '../src/lib.js';
import localStorage from './localStorage.js';

describe('JSONDate', function() {
  describe('Date.prototype.toJSON()', function() {
    it('should return a string representation of the Date object', function() {
      const event = new Date(1995, 11, 17, 3, 24, 0);
      const jsonDate = event.toJSON();
      assert.equal(jsonDate, "1995-12-17T02:24:00.000Z");
    });
  });

  describe('now', function() {
    it('should return a jsonDate string', function() {
      assert.ok(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/.test(JSONDate.now()));
    });
  });

  describe('getDate', function() {
    it('should return the yyyy-mm-dd part of the jsonDate string', function() {
      assert.ok(/9999-99-99/.test(JSONDate.getDate("9999-99-99")));
      assert.ok(/^\d{4}-\d{2}-\d{2}$/.test(JSONDate.getDate("1995-12-17T02:24:00.000Z")));
      assert.ok(/^\d{4}-\d{2}-\d{2}$/.test(JSONDate.getDate("2000-01-02T12:02:00.000Z")));
      assert.ok(/1995-12-17/.test(JSONDate.getDate("1995-12-17T02:24:00.000Z")));
      assert.ok(/2000-01-02/.test(JSONDate.getDate("2000-01-02T12:02:00.000Z")));
    });
  });

  describe('getTime', function() {
    it('should return the hh:mm:ss part of the jsonDate string', function() {
      assert.ok(/00:00:00/.test(JSONDate.getTime("9999-99-99X00:00:00.XXXXX")));
      assert.ok(/^\d{2}:\d{2}:\d{2}$/.test(JSONDate.getTime("1995-12-17T02:24:00.000Z")));
      assert.ok(/^\d{2}:\d{2}:\d{2}$/.test(JSONDate.getTime("2000-01-02T12:02:00.000Z")));
      assert.ok(/02:24:00/.test(JSONDate.getTime("1995-12-17T02:24:00.000Z")));
      assert.ok(/12:02:00/.test(JSONDate.getTime("2000-01-02T12:02:00.000Z")));
    });
  });
});

describe('Counter', function() {
  describe('count', function() {
    it('should count the frequency of each element (number)', function() {
      const array = [1, 2, 1, 3];
      const freq = Counter.count(array);
      assert.equal(freq[1], 2);
      assert.equal(freq[2], 1);
      assert.equal(freq[3], 1);
      assert.equal(freq[4], undefined);
    });

    it('should count the frequency of each element (string)', function() {
      const array = ['1', '2', '1', '3'];
      const freq = Counter.count(array);
      assert.equal(freq[1], 2);
      assert.equal(freq[2], 1);
      assert.equal(freq[3], 1);
      assert.equal(freq[4], undefined);
    });
  });
});

describe('DateCounter', function() {
  describe('count', function() {
    it('should count the frequency of each date string (yyyy-mm-dd)', function() {
      const array = ['2001-01-01', '1995-01-10', '1991-10-02', '1995-01-10'];
      const freq = Counter.count(array);
      assert.equal(freq['2001-01-01'], 1);
      assert.equal(freq['1995-01-10'], 2);
      assert.equal(freq['1991-10-02'], 1);
      assert.equal(freq['1990-10-02'], undefined);
    });

    it('should count the frequency of each JSONDate string (0000-00-00T00:00:00.000Z)', function() {
      const array = [
        '2001-01-01T02:24:00.000Z', 
        '1995-01-10T02:24:00.000Z', 
        '1991-10-02T02:24:00.000Z', 
        '1995-01-10T13:00:12.000Z'
      ];
      const freq = DateCounter.count(array);
      assert.equal(freq['2001-01-01'], 1);
      assert.equal(freq['1995-01-10'], 2);
      assert.equal(freq['1991-10-02'], 1);
      assert.equal(freq['1990-10-02'], undefined);
    });
  });
});

describe('Logger', function() {
  describe('log', function() {
    it('should add one element to the beginning of an array', function() {
      const logger = new Logger(localStorage);
      // inserting 1
      logger.log(1);
      assert.deepEqual(logger.getLogs(), [1]);
      // inserting 2
      logger.log(2);
      assert.deepEqual(logger.getLogs(), [2, 1]);
    });
  });
});

/*
describe('', function() {
  it('should ', function() {

  });
});
*/

test/localStorage.js
if (typeof localStorage === 'undefined') {
  global.localStorage = {
    storage: {},
    setItem: function(key, value) {
      this.storage[key] = value;
    },
    getItem: function(key) {
      return this.storage[key];
    }
  }
}

export default localStorage;

src/app.js
import { DateCounter, JSONDate, Logger } from './lib.js';

function renderLogs(logs, $el) {
  let html = '';
  logs.forEach(jsonDate => {
    const dateTime = `${JSONDate.getDate(jsonDate)}, ${JSONDate.getTime(jsonDate)}`;
    html += `<li>${dateTime}</li>`;
  });
  $el.innerHTML = html;
}

function renderDailyStats(stats, $el) {
  let html = '';
  for (let date in stats) {
      html += `<tr><td>${date}</td><td>${stats[date]}</td></tr>`;
  }
  $el.innerHTML = html;
}

function render() {
  const logs = logger.getLogs();
  const stats = DateCounter.count(logs);
  renderLogs(logs, $logs);
  renderDailyStats(stats, $stats);
  $copyText.value = JSON.stringify(logs);
  $saveText.value = "";
}

function log() {
  logger.log(JSONDate.now());
  render();
}

function copyText() {
  $copyText.select();
  $copyText.setSelectionRange(0, 99999); // For mobile devices
  // Copy the text inside the text field 
  document.execCommand("copy");
}

function saveLogs() {
  logger.import($saveText.value);
  render();
}

document.addEventListener('DOMContentLoaded', () => {
  if(typeof(Storage) !== 'undefined') {
    logger = new Logger(localStorage);
    $logs = document.getElementById('logs');
    $stats = document.getElementById('stats');
    $logBtn = document.getElementById("logBtn");
    $copyBtn = document.getElementById("copyBtn");
    $copyText = document.getElementById("copyText");
    $saveBtn = document.getElementById("saveBtn");
    $saveText = document.getElementById("saveText");
    
    $logBtn.addEventListener('click', log);
    $copyBtn.addEventListener('click', copyText);
    $saveBtn.addEventListener('click', saveLogs);
    
    render();
  } else {
    alert('Sorry, your browser does not support web storage...');
  }
}, false);

let logger, $logs, $stats, $logBtn, $copyBtn, $copyText, $saveBtn, $saveText;

src/lib.js
const noop = () => {};

class JSONDate {
  // return a string like "2000-12-17T02:24:00.000Z"
  static now() {
    const date = new Date();
    // local time adjustment
    date.setMinutes(date.getMinutes() - date.getTimezoneOffset());
    const jsonDate = date.toJSON(); // UTC 
    return jsonDate;
  }

  // Given "2000-12-17T02:24:00.000Z" return "2000-12-17"
  static getDate(jsonDate) {
    return jsonDate.substring(0, 10);
  }

  // Given "2000-12-17T02:24:00.000Z" return "02:24:00"
  static getTime(jsonDate) {
    return jsonDate.substring(11, 19);
  }
}

class Counter {
  /*
  * Given the array [ 1, 2, 1, 3 ]
  * it returns { '1': 2, '2': 1, '3': 1 }
  */
  static count(array) {
    return array.reduce((map, el) => {
      map[el] = map[el] ? (map[el] + 1) : 1;
      return map;
    }, {});
  }
}

class DateCounter extends Counter {
  static count(array) {
    return Counter.count(array.map(jsonDate => JSONDate.getDate(jsonDate)));
  }
}

class Logger {
  constructor(storage, id = 'log') {
    this.storage = storage;
    this.id = id;
  }

  getLogs() {
    const array = this.storage.getItem(this.id);
    return array && JSON.parse(array) || [];
  }

  log(value) { //, onUpdated = noop) {
    const logs = this.getLogs();
    logs.unshift(value); // prepend the value
    this.storage.setItem(this.id, JSON.stringify(logs));
    //onUpdated();
  }

  import(logs) {
    this.storage.setItem(this.id, logs);
  }
}

export { 
  JSONDate, 
  Counter, 
  DateCounter, 
  Logger 
}; 

script.js
(function() {

class ObjStorage {
  static set(key, value) {
    localStorage.setItem(key, JSON.stringify(value));
  }

  static get(key) {
    var value = localStorage.getItem(key);
    return value && JSON.parse(value);
  }

  static test() {
    ObjStorage.set('test', { 'one': 1, 'two': 2, 'three': 3 });
    console.log(ObjStorage.get('test'));
  }
}

class JSONDate {
  static now() {
    var date = new Date();
    //date.setDate(31);
    var jsonDate = date.toJSON(); // UTC
    return jsonDate;
  }

  static backToDate(jsonDate) {
    var backToDate = new Date(jsonDate);
    return backToDate;
  }

  static getDate(jsonDate) {
    //return jsonDate.substring(0, 10);
    var date = JSONDate.backToDate(jsonDate);
    //return date.toLocaleDateString();
    var yyyy = date.getFullYear();
    var mm = ((date.getMonth() + 1) < 10 ? '0' : '') + (date.getMonth() + 1);
    var dd = (date.getDate() < 10 ? '0' : '') + date.getDate();
    return yyyy + '-' + mm + '-' + dd;
  }

  static getTime(jsonDate) {
    //return jsonDate.substring(11, 19);
    var date = JSONDate.backToDate(jsonDate);
    return date.toLocaleTimeString();
  }

  static test() {
    var jsonDate = JSONDate.now()
    console.log(Date())
    console.log(jsonDate)
    console.log(JSONDate.backToDate(jsonDate))
  }
}

class Logger {    
  static logs(key) {
    var logs = ObjStorage.get(key);
    return logs || [];
  }

  static log(key, cb) {
    var logs = Logger.logs(key);
    var date = JSONDate.now();

    // prepend the date
    logs.unshift(date);

    ObjStorage.set(key, logs);
  
    cb(logs);
  }
  
  static dailyStats(logs) {
    var stats = {};
    
    // group logs by date (aaaa-mm-dd)
    stats = logs.reduce( (stats, curr) => {
      var date = JSONDate.getDate(curr);
      stats[date] = stats[date] ? (stats[date] + 1) : 1;
      return stats;
    }, stats);

    return stats;
  }
}

function renderLogs(logs, $el) {
  var html = '';

  logs.forEach(jsonDate => {
    //console.log(jsonDate);
    //var date = JSONDate.backToDate(jsonDate);
    //var dateTime = date.toLocaleString();
    var dateTime = JSONDate.getDate(jsonDate) + ', ' + JSONDate.getTime(jsonDate);
    
    html += '<li>' + dateTime + '</li>';
  });
  
  $el.innerHTML = html;
}

function renderDailyStats(logs, $el) {
  var html = '';

  var stats = Logger.dailyStats(logs);

  for (let date in stats) {
    //console.log(date); 
    html += '<tr>';
    html += '<td>' + date + '</td>';
    html += '<td>' + stats[date] + '</td>';
    html += '</tr>';
  }

  $el.innerHTML = html;
}

function render(props={}) {
  var logs = props.logs || Logger.logs(_log);
  //renderClickCount(props.clickcount, $clickCount);
  renderLogs(logs, $logs);
  renderDailyStats(logs, $stats);
}

function log() {
  Logger.log(_log, (logs) => {
    render({
      'logs': logs,
    });
  });
}

function init() {
  if(typeof(Storage) !== 'undefined') {
    $btnLog = document.getElementById('btnLog');
    $logs = document.getElementById('logs');
    $stats = document.getElementById('stats');

    $btnLog.addEventListener('click', log);

    render();

  } else {
    alert('Sorry, your browser does not support web storage...');
  }
}

function _init() {
  _log = 'log'; //LogTypes.select(0);
  document.addEventListener('DOMContentLoaded', init, false);
}

var $btnLog, $logs, $stats, _log;

_init();

})();